engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java

```diff
@@ -68,32 +70,44 @@ public final Object findCachedResult(UUID nodeId) {
     // item in the `functionCallCallback`. We allow to execute the cached `stackTop` value to be
     // able to continue the stack execution, and unwind later from the `onReturnValue` callback.
     if (result != null && !nodeId.equals(nextExecutionItem)) {
-      var value = new ExpressionValue(
+      var value =
+          new ExpressionValue(
               nodeId,
               result,
               cache.getType(nodeId),
               typeOf(result),
               calls.get(nodeId),
               cache.getCall(nodeId),
-              new ProfilingInfo[]{ExecutionTime.empty()},
-              true
-      );
+              new ProfilingInfo[] {ExecutionTime.empty()},
+              true);
       onCachedCallback.accept(value);
       return result;
     }
     return null;
   }
 
   @CompilerDirectives.TruffleBoundary
-  public final void updateCachedResult(UUID nodeId, Object result, boolean isPanic, long nanoTimeElapsed) {
+  public final void updateCachedResult(
+      Object node, Object result, boolean isPanic, long nanoTimeElapsed) {
+    ExpressionNode expressionNode = (ExpressionNode) node;
+    UUID nodeId = expressionNode.getId();
     String resultType = typeOf(result);
     String cachedType = cache.getType(nodeId);
     FunctionCallInfo call = functionCallInfoById(nodeId);
     FunctionCallInfo cachedCall = cache.getCall(nodeId);
-    ProfilingInfo[] profilingInfo = new ProfilingInfo[]{new ExecutionTime(nanoTimeElapsed)};
+    ProfilingInfo[] profilingInfo = new ProfilingInfo[] {new ExecutionTime(nanoTimeElapsed)};
 
-    ExpressionValue expressionValue
-            = new ExpressionValue(nodeId, result, resultType, cachedType, call, cachedCall, profilingInfo, false);
+    ExpressionValue expressionValue =
+        new ExpressionValue(
+            nodeId,
+            result,
+            resultType,
+            cachedType,
+            call,
+            cachedCall,
+            profilingInfo,
+            false,
+            expressionNode);
```

Comment:
I capture the node here to get its local scope

Reply:
`Node` is program. You are creating a _value_ which points to a program. That's very unusual and probably not correct.